File: src/repo_config.py
import subprocess
import re
import os
import gettext

script_dir = os.path.dirname(os.path.abspath(__file__))
localedir = os.path.join(os.path.dirname(script_dir), "locale")
if os.path.exists(localedir):
    try:
        t = gettext.translation("mirrorman", localedir=localedir, fallback=True)
        _ = t.gettext
    except Exception:

        def _(s):
            return s
else:
    try:
        t = gettext.translation(
            "mirrorman", localedir="/usr/share/locale", fallback=True
        )
        _ = t.gettext
    except Exception:

        def _(s):
            return s


def get_suitable_terminal():
    """Detect and return the command to open a terminal based on desktop environment."""
    desktop = os.environ.get("XDG_CURRENT_DESKTOP", "").upper()
    if "GNOME" in desktop:
        if subprocess.run(["which", "ptyxis"], capture_output=True).returncode == 0:
            return "ptyxis -x "
        elif (
            subprocess.run(["which", "gnome-terminal"], capture_output=True).returncode
            == 0
        ):
            return "gnome-terminal -- "
    elif "KDE" in desktop or "PLASMA" in desktop:
        if subprocess.run(["which", "konsole"], capture_output=True).returncode == 0:
            return "konsole -e "
    elif "XFCE" in desktop:
        if (
            subprocess.run(["which", "xfce4-terminal"], capture_output=True).returncode
            == 0
        ):
            return "xfce4-terminal -e "
    for term in ["alacritty", "kitty", "xterm"]:
        if subprocess.run(["which", term], capture_output=True).returncode == 0:
            return f"{term} -e "
    return None


def open_terminal_with_repo_command(command, parent_window=None):
    """Open a terminal with the given command."""
    escaped_cmd = command.replace("'", "'\"'\"'")
    terminal_cmd = get_suitable_terminal()
    if terminal_cmd is None:
        if parent_window:
            parent_window.show_error_dialog(
                _("Terminal Error"), _("No suitable terminal emulator found.")
            )
        else:
            print("Error: No suitable terminal emulator found.")
        return
    full_cmd = f"{terminal_cmd} bash -c '{escaped_cmd}; echo; echo Press ENTER to close...; read'"
    try:
        subprocess.Popen(full_cmd, shell=True)
    except Exception as e:
        if parent_window:
            parent_window.show_error_dialog(_("Terminal Error"), str(e))
        else:
            print(f"Error opening terminal: {e}")


class RepoConfig:
    def __init__(self):
        self.pacman_conf = "/etc/pacman.conf"
        self.standard_repos = ["core", "extra", "multilib"]
        self.third_party_repos = ["chaotic-aur", "blackarch", "archlinuxcn"]
        self.repositories = {
            repo: False for repo in self.standard_repos + self.third_party_repos
        }
        self.load_pacman_conf()

    def load_pacman_conf(self):
        """Load and parse pacman.conf to determine enabled repositories."""
        try:
            with open(self.pacman_conf, "r") as f:
                lines = f.readlines()
            repo_pattern = re.compile(r"^\s*(#?)\s*\[([^]]+)\]\s*$")
            current_repo = None
            for line in lines:
                match = repo_pattern.match(line)
                if match:
                    comment, repo_name = match.groups()
                    if repo_name == "options":
                        continue
                    current_repo = repo_name
                    enabled = not bool(comment)
                    if repo_name in self.repositories:
                        self.repositories[repo_name] = enabled
        except Exception as e:
            print(f"Error loading pacman.conf: {e}")

    def get_repositories(self):
        """Return the dictionary of repositories and their enabled status."""
        return self.repositories

    def set_repository_enabled(self, repo_name, enabled, parent_window=None):
        """Enable or disable a repository."""
        if repo_name not in self.repositories:
            print(f"Repository {repo_name} not supported.")
            return
        current_enabled = self.repositories[repo_name]
        if current_enabled == enabled:
            return  # Already in desired state
        if repo_name in self.standard_repos:
            if enabled:
                sed_cmd = f"sudo sed -i '/^#\\[{repo_name}\\]/,/^\\[/s/^#//g' {self.pacman_conf}"
            else:
                sed_cmd = f"sudo sed -i '/^\\[{repo_name}\\]/,/^\\[/s/^/#/' {self.pacman_conf}"
            commands = [sed_cmd, "sudo pacman -Syy"]
            cmd_str = " && ".join(commands)
            open_terminal_with_repo_command(cmd_str, parent_window)
        elif repo_name in self.third_party_repos:
            func_name = repo_name.replace("-", "_")
            if enabled:
                getattr(self, f"enable_{func_name}")(parent_window)
            else:
                getattr(self, f"disable_{func_name}")(parent_window)

    def add_repository(self, repo_name, repo_url, parent_window=None):
        """Add a new repository to pacman.conf."""
        if repo_name in self.repositories:
            print(f"Repository {repo_name} already exists or is reserved.")
            return
        commands = [
            "echo '' | sudo tee -a /etc/pacman.conf",
            f"echo '[{repo_name}]' | sudo tee -a /etc/pacman.conf",
            f"echo 'Server = {repo_url}' | sudo tee -a /etc/pacman.conf",
            "sudo pacman -Syy",
        ]
        cmd_str = " && ".join(commands)
        open_terminal_with_repo_command(cmd_str, parent_window)
        self.repositories[repo_name] = True

    def enable_chaotic_aur(self, parent_window=None):
        """Enable Chaotic-AUR repository."""
        commands = [
            "sudo pacman-key --recv-key 3056513887B78AEB --keyserver keyserver.ubuntu.com",
            "sudo pacman-key --lsign-key 3056513887B78AEB",
            "sudo pacman -U 'https://cdn-mirror.chaotic.cx/chaotic-aur/chaotic-keyring.pkg.tar.zst' --noconfirm",
            "sudo pacman -U 'https://cdn-mirror.chaotic.cx/chaotic-aur/chaotic-mirrorlist.pkg.tar.zst' --noconfirm",
            "echo '' | sudo tee -a /etc/pacman.conf",
            "echo '[chaotic-aur]' | sudo tee -a /etc/pacman.conf",
            "echo 'Include = /etc/pacman.d/chaotic-mirrorlist' | sudo tee -a /etc/pacman.conf",
            "sudo pacman -Syy",
        ]
        cmd_str = " && ".join(commands)
        open_terminal_with_repo_command(cmd_str, parent_window)

    def disable_chaotic_aur(self, parent_window=None):
        """Disable Chaotic-AUR repository."""
        commands = [
            # Remove the entire [chaotic-aur] section until the next section or end of file
            r"sudo sed -i '/\[chaotic-aur\]/,/^\[/{/^\[/!d; /^\[chaotic-aur\]/d}' /etc/pacman.conf",
            "sudo pacman -Syy",
        ]
        cmd_str = " && ".join(commands)
        open_terminal_with_repo_command(cmd_str, parent_window)

    def enable_blackarch(self, parent_window=None):
        """Enable BlackArch repository."""
        commands = [
            "cd /tmp",
            "curl -O https://blackarch.org/strap.sh",
            "echo '26849980b35a42e6e192c6d9ed8c46f0d6d06047 strap.sh' | sha1sum -c",
            "if [ $? -eq 0 ]; then chmod +x strap.sh && sudo ./strap.sh; else echo 'SHA1 verification failed!'; exit 1; fi",
            "rm -f strap.sh",
            "sudo pacman -Syy",
        ]
        cmd_str = " && ".join(commands)
        open_terminal_with_repo_command(cmd_str, parent_window)

    def disable_blackarch(self, parent_window=None):
        """Disable BlackArch repository."""
        commands = [
            # Remove the entire [blackarch] section until the next section or end of file
            r"sudo sed -i '/\[blackarch\]/,/^\[/{/^\[/!d; /^\[blackarch\]/d}' /etc/pacman.conf",
            "sudo pacman -Syy",
        ]
        cmd_str = " && ".join(commands)
        open_terminal_with_repo_command(cmd_str, parent_window)

    def enable_archlinuxcn(self, parent_window=None):
        """Enable ArchLinuxCN repository."""
        commands = [
            "echo '' | sudo tee -a /etc/pacman.conf",
            "echo '[archlinuxcn]' | sudo tee -a /etc/pacman.conf",
            "echo 'Server = https://repo.archlinuxcn.org/$arch' | sudo tee -a /etc/pacman.conf",
            "sudo pacman -Syy",
            "sudo pacman -S archlinuxcn-keyring --noconfirm",
        ]
        cmd_str = " && ".join(commands)
        open_terminal_with_repo_command(cmd_str, parent_window)

    def disable_archlinuxcn(self, parent_window=None):
        """Disable ArchLinuxCN repository."""
        commands = [
            # Remove the entire [archlinuxcn] section until the next section or end of file
            r"sudo sed -i '/\[archlinuxcn\]/,/^\[/{/^\[/!d; /^\[archlinuxcn\]/d}' /etc/pacman.conf",
            "sudo pacman -Syy",
        ]
        cmd_str = " && ".join(commands)
        open_terminal_with_repo_command(cmd_str, parent_window)

File: src/repo_config.rs
use std::collections::HashMap;

const PACMAN_CONF: &str = "/etc/pacman.conf";

pub struct RepoConfig {
    pub pacman_conf: &'static str,
    pub standard_repos: Vec<String>,
    pub third_party_repos: Vec<String>,
    pub custom_repos: Vec<String>,
    pub repositories: HashMap<String, bool>,
}

impl RepoConfig {
    pub fn new() -> Self {
        let standard_repos = vec![
            "core".to_string(),
            "extra".to_string(),
            "multilib".to_string(),
        ];
        let third_party_repos = vec![
            "chaotic-aur".to_string(),
            "blackarch".to_string(),
            "archlinuxcn".to_string(),
        ];

        let mut repositories = HashMap::new();
        for repo in standard_repos.iter().chain(third_party_repos.iter()) {
            repositories.insert(repo.clone(), false);
        }

        let mut config = Self {
            pacman_conf: PACMAN_CONF,
            standard_repos,
            third_party_repos,
            custom_repos: Vec::new(),
            repositories,
        };

        config.load_pacman_conf();
        config
    }

    fn load_pacman_conf(&mut self) {
        let content = match std::fs::read_to_string(self.pacman_conf) {
            Ok(c) => c,
            Err(_) => return,
        };

        let repo_pattern = regex_lite::Regex::new(r"^\s*(#?)\s*\[([^\]]+)\]").unwrap();

        for line in content.lines() {
            if let Some(caps) = repo_pattern.captures(line) {
                let repo_name = caps.get(2).unwrap().as_str().to_string();
                if repo_name == "options" {
                    continue;
                }
                let is_commented = caps.get(1).unwrap().as_str() == "#";
                let enabled = !is_commented;

                if self.repositories.contains_key(&repo_name) {
                    self.repositories.insert(repo_name, enabled);
                }
            }
        }
    }

    pub fn toggle_repo_in_config(
        &mut self,
        repo_name: &str,
        enable: bool,
        is_third_party: bool,
    ) -> Result<(), String> {

        let config_text =
            std::fs::read_to_string(self.pacman_conf).map_err(|e| format!("Failed to read pacman.conf: {e}"))?;

        let section_snippet = if is_third_party {
            get_third_party_section(repo_name)
        } else {
            None
        };

        let modified = toggle_repo_text(&config_text, repo_name, enable, section_snippet.as_deref());
        crate::helper_client::HelperClient::save_pacman_conf(&modified)?;

        self.repositories.insert(repo_name.to_string(), enable);
        Ok(())
    }

    pub fn add_repository(&mut self, repo_name: &str, repo_url: &str, siglevel: &str) -> Result<(), String> {
        if repo_name.is_empty() || repo_url.is_empty() {
            return Err("Repository name and URL are required".to_string());
        }
        if self.repositories.contains_key(repo_name) {
            return Err(format!("Repository already exists: '{repo_name}'"));
        }

        let config_text = if let Ok(c) = std::fs::read_to_string(self.pacman_conf) {
            c
        } else {
            String::new()
        };

        let sig_line = if siglevel.is_empty() {
            String::new()
        } else {
            format!("SigLevel = {siglevel}\n")
        };
        let modified = format!("{config_text}\n[{repo_name}]\nServer = {repo_url}\n{sig_line}");
        crate::helper_client::HelperClient::save_pacman_conf(&modified)?;

        self.repositories.insert(repo_name.to_string(), true);
        self.custom_repos.push(repo_name.to_string());
        Ok(())
    }

    pub fn enable_third_party(&self, repo_name: &str) -> Result<(), String> {
        match repo_name {
            "chaotic-aur" => enable_chaotic_aur(),
            "blackarch" => enable_blackarch(),
            "archlinuxcn" => enable_archlinuxcn(),
            _ => Err(format!("Unknown third-party repo: {repo_name}")),
        }
    }
}

fn toggle_repo_text(config_text: &str, repo_name: &str, enable: bool, section_snippet: Option<&str>) -> String {
    let mut new_lines = Vec::new();
    let mut found_section = false;
    let mut in_section = false;
    let section_header = format!("[{repo_name}]");
    let snippet_lines: Vec<&str> = section_snippet.map(|s| s.lines().collect()).unwrap_or_default();

    for line in config_text.lines() {
        let stripped = line.trim();
        let header_check = stripped.trim_start_matches('#').trim();
        if header_check == section_header {
            found_section = true;
            in_section = true;
            if enable {
                new_lines.push(line.trim_start_matches('#').to_string());
                // Add replacement lines from snippet
                for sl in &snippet_lines[1..] {
                    if sl.starts_with("Include =") || sl.starts_with("Server =") || sl.starts_with("SigLevel =") {
                        new_lines.push(sl.to_string());
                    }
                }
            } else {
                if !line.starts_with('#') {
                    new_lines.push(format!("#{line}"));
                } else {
                    new_lines.push(line.to_string());
                }
            }
            continue;
        }

        if in_section {
            let header_uncommented = stripped.trim_start_matches('#').trim();
            if header_uncommented.starts_with('[') && header_uncommented != section_header {
                in_section = false;
                new_lines.push(line.to_string());
                continue;
            }
            if header_uncommented.starts_with("Include =") || header_uncommented.starts_with("Server =") || header_uncommented.starts_with("SigLevel =") {
                if !enable {
                    if !line.starts_with('#') {
                        new_lines.push(format!("#{line}"));
                    } else {
                        new_lines.push(line.to_string());
                    }
                } else if snippet_lines.is_empty() {
                    new_lines.push(line.trim_start_matches('#').to_string());
                }
                // When enabling with snippet: skip old line (replaced above)
                // When enabling without snippet (standard repos): keep existing line
            } else {
                new_lines.push(line.to_string());
            }
        } else {
            new_lines.push(line.to_string());
        }
    }

    if enable && !found_section {
        if let Some(snippet) = section_snippet {
            new_lines.push(String::new());
            for snippet_line in snippet.lines() {
                new_lines.push(snippet_line.to_string());
            }
        }
    }

    new_lines.push(String::new());
    new_lines.join("\n")
}

fn get_third_party_section(repo_name: &str) -> Option<String> {
    match repo_name {
        "chaotic-aur" => Some("[chaotic-aur]\nSigLevel = Optional TrustAll\nInclude = /etc/pacman.d/chaotic-mirrorlist\n".to_string()),
        "blackarch" => Some(
            "[blackarch]\nSigLevel = Optional\nServer = https://blackarch.org/blackarch/$repo/os/$arch\n".to_string(),
        ),
        "archlinuxcn" => {
            Some("[archlinuxcn]\nSigLevel = Optional TrustAll\nServer = https://repo.archlinuxcn.org/$arch\n".to_string())
        }
        _ => None,
    }
}

fn enable_chaotic_aur() -> Result<(), String> {
    run_cmd(
        "pacman-key",
        &[
            "--recv-key",
            "3056513887B78AEB",
            "--keyserver",
            "keyserver.ubuntu.com",
        ],
    )?;
    run_cmd("pacman-key", &["--lsign-key", "3056513887B78AEB"])?;
    run_cmd(
        "pacman",
        &[
            "-U",
            "--noconfirm",
            "https://cdn-mirror.chaotic.cx/chaotic-aur/chaotic-keyring.pkg.tar.zst",
            "https://cdn-mirror.chaotic.cx/chaotic-aur/chaotic-mirrorlist.pkg.tar.zst",
        ],
    )?;
    Ok(())
}

fn enable_blackarch() -> Result<(), String> {
    run_cmd(
        "bash",
        &[
            "-c",
            "cd /tmp && curl -O https://blackarch.org/strap.sh && echo '26849980b35a42e6e192c6d9ed8c46f0d6d06047  strap.sh' | sha1sum -c && chmod +x strap.sh && ./strap.sh && rm -f strap.sh",
        ],
    )?;
    Ok(())
}

fn enable_archlinuxcn() -> Result<(), String> {
    run_cmd(
        "pacman-key",
        &[
            "--recv-key",
            "4D41FD3D9E72E7966A573093E8CA6AEB220E236C",
            "--keyserver",
            "keyserver.ubuntu.com",
        ],
    )?;
    run_cmd(
        "pacman-key",
        &["--lsign-key", "4D41FD3D9E72E7966A573093E8CA6AEB220E236C"],
    )?;
    run_cmd("pacman", &["-S", "archlinuxcn-keyring", "--noconfirm"])?;
    Ok(())
}

fn run_cmd(command: &str, args: &[&str]) -> Result<(), String> {
    if command == "pacman" {
        let (success, _, stderr) = crate::helper_client::HelperClient::run_pacman(args)?;
        if success {
            Ok(())
        } else {
            Err(format!("pacman failed: {stderr}"))
        }
    } else {
        let (success, _, stderr) = crate::helper_client::HelperClient::run_command(command, args)?;
        if success {
            Ok(())
        } else {
            Err(format!("Command '{command}' failed: {stderr}"))
        }
    }
}
